### PR TITLE
[D152] 

### PR DESCRIPTION
**Pull request by tienhoah**
**Features:**
- Dynamically jump to cluster when loads.
- Full zoom now responds to initial position of the clusters.
- Improved presentation of pie chart.
- Improved componentDidUpdate in leaflet map component.

**Bug Fix:**
- Handle updateBoundingBoxFromPan when panning at the same location.
- Increased padding when using leaflet getBounds with marker cluster to prevent missing covered data.

### CODE REVIEW

File: src/components/LeafletMap.js [hunk 1]
```diff
@@ -309,8 +307,8 @@ class LeafletMap extends Component {
       const currBounds = e.target.getBounds();
       const currZoom = e.target.getZoom();
       const currBbox = this.createBoundingBox(currBounds);
-      const zoomDiff = mapZoomLevel - currZoom;
-      if (zoomDiff !== 0) {
+
+      if (mapZoomLevel !== null) {
```
Comment (brianbancroft) [inline src/components/LeafletMap.js]: is there a point where this might be undefined?

File: src/components/LeafletMap.js [hunk 2]
```diff
@@ -326,12 +324,18 @@ class LeafletMap extends Component {
       const currBounds = e.target.getBounds();
       const southBound = currBounds.getSouth();
       const currBbox = this.createBoundingBox(currBounds);
-      const viewDiff = mapBoundingBoxCorner.toFixed(5) - southBound.toFixed(5);
-      if (viewDiff !== 0) {
-        this.setState({ mapBoundingBoxCorner: southBound });
-        getMapBounds(currBbox);
+
+      if (mapBoundingBoxCorner !== null) {
```
Comment (brianbancroft) [inline src/components/LeafletMap.js]: is there a point where this might be undefined?

File: src/components/LeafletMap.js [hunk 3]
```diff
@@ -326,12 +324,18 @@ class LeafletMap extends Component {
...
+        const viewDiff = mapBoundingBoxCorner - southBound;
+        if (viewDiff === 0) {
+          this.setState({ panToSameView: true });
+          contentLoadHandler(false);
+        } else if (viewDiff !== 0) {
```
Comment (brianbancroft) [inline src/components/LeafletMap.js]: is the if in the else-if redundant here?

File: src/components/LeafletMap.js [hunk 4]
```diff
@@ -504,6 +509,24 @@ class LeafletMap extends Component {
     });
 
     this.map.addLayer(markersCluster);
+
+    if (markersData.length !== 0) {
```
Comment (brianbancroft) [inline src/components/LeafletMap.js]: An explicit condition is redundant here. 

image

Comment (tienhoah) [reply src/components/LeafletMap.js]: there is case when no markers or when sites data are not returned, so the map is set to default

Comment (brianbancroft) [reply src/components/LeafletMap.js]: So I see that you're now using `if (markersData.length > 0) { ...` here in place. 

You don't really need that either. But I'm going to approve all the same.

**Comment by tienhoah** [comment]
closes #152

**Pull request merged**